system/lib/wasmfs/wasmfs.cpp

```diff
@@ -62,9 +62,46 @@ extern "C" void wasmfs_flush(void) {
   // Flush musl libc streams.
   fflush(0);
 
-  // Flush our own streams. TODO: flush all backends.
+  // Flush our own streams.
   (void)SpecialFiles::getStdout()->locked().flush();
   (void)SpecialFiles::getStderr()->locked().flush();
+
+  std::vector<std::shared_ptr<Directory>> toFlush;
+  toFlush.push_back(wasmFS.getRootDirectory());
+
+  while (!toFlush.empty()) {
+    auto dir = toFlush.back();
+    toFlush.pop_back();
+
+    auto lockedDir = dir->locked();
+    Directory::MaybeEntries entries = lockedDir.getEntries();
+    if (int err = entries.getError()) {
+#ifndef NDEBUG
+      std::string errorMessage =
+        "Non-fatal error code " + std::to_string(err) +
+        " while flushing directory: " + lockedDir.getName(dir);
+      emscripten_console_error(errorMessage.c_str());
+#endif
+      continue;
+    }
+    for (auto entry : *entries) {
+      if (entry.kind == File::FileKind::DataFileKind) {
+        if (int err = lockedDir.getChild(entry.name)
+                        ->dynCast<DataFile>()
+                        ->locked()
+                        .flush()) {
+#ifndef NDEBUG
+          std::string errorMessage = "Non-fatal error code " +
+                                     std::to_string(err) +
+                                     " while flushing file: " + entry.name;
+          emscripten_console_error(errorMessage.c_str());
+#endif
+        }
+      } else if (entry.kind == File::FileKind::DirectoryKind) {
+        toFlush.push_back(lockedDir.getChild(entry.name)->dynCast<Directory>());
```

Comment:
It would be good to do the `getChild` and the `dynCast` in the condition rather than trusting the `entry.kind`. This will protect against bugs caused by different threads deleting a directory and creating a different kind of file with the same name in between when we called `getEntries` and when we call `getChild`. Similarly, we should handle the case where the child no longer exists.

Reply:
Yes and yes. If a child has been deleted on a different thread so the sync code gets a null from `getChild`, then I think we can just ignore it and not return an error, even in a debug build.

Reply:
If we decide to ignore it, is a check needed for the null case at all? i.e. wouldn't checking if `lockedDir.getChild(entry.name)->dynCast<Directory>()` and `else if` for `lockedDir.getChild(entry.name)->dynCast<DataFile>()` catch all the valid cases, leaving the null case to not trigger any code and just be ignored?

Reply:
Almost, except that calling `dynCast` on a null pointer would be undefined behavior :)